### PR TITLE
CAM: use IPCHelper

### PR DESCRIPTION
**Pull request by wwylele**
Please keep eyes on those numbers to see if I made mistakes, thanks!

---
This change is [Reviewable](https://reviewable.io/reviews/citra-emu/citra/2722)

### CODE REVIEW

File: src/core/hle/service/cam/cam.cpp [hunk 1]
```diff
@@ -403,287 +395,272 @@ void SetReceiving(Service::Interface* self) {
             port.is_pending_receiving = true;
         }
 
-        cmd_buff[1] = RESULT_SUCCESS.raw;
-        cmd_buff[2] = IPC::CopyHandleDesc();
-        cmd_buff[3] = Kernel::g_handle_table.Create(port.completion_event).MoveFrom();
+        rb.Push(RESULT_SUCCESS);
+        rb.PushCopyHandles(Kernel::g_handle_table.Create(port.completion_event).MoveFrom());
     } else {
-        LOG_ERROR(Service_CAM, "invalid port_select=%u", port_select.m_val);
```
Comment (Rinkaa) [inline src/core/hle/service/cam/cam.cpp]: Intentionally removed?

File: src/core/hle/service/cam/cam.cpp [hunk 2]
```diff
@@ -403,287 +395,272 @@ void SetReceiving(Service::Interface* self) {
...
+    IPC::RequestParser rp(Kernel::GetCommandBuffer(), 0x08, 1, 0);
+    const PortSet port_select(rp.Pop<u8>());
 
+    IPC::RequestBuilder rb = rp.MakeBuilder(2, 0);
     if (port_select.IsSingle()) {
         int port = *port_select.begin();
-        cmd_buff[1] = RESULT_SUCCESS.raw;
-        cmd_buff[2] = (ports[port].is_receiving || ports[port].is_pending_receiving) ? 0 : 1;
+        rb.Push(RESULT_SUCCESS);
+        rb.Push(ports[port].is_receiving || ports[port].is_pending_receiving);
```
Comment (Rinkaa) [inline src/core/hle/service/cam/cam.cpp]: Intentional?

File: src/core/hle/service/cam/cam.cpp [hunk 3]
```diff
@@ -403,287 +395,272 @@ void SetReceiving(Service::Interface* self) {
...
-    const PortSet port_select(cmd_buff[1]);
+    IPC::RequestParser rp(Kernel::GetCommandBuffer(), 0x0C, 1, 0);
+    const PortSet port_select(rp.Pop<u8>());
 
+    IPC::RequestBuilder rb = rp.MakeBuilder(2, 0);
     if (port_select.IsSingle()) {
         int port = *port_select.begin();
-        cmd_buff[1] = RESULT_SUCCESS.raw;
-        cmd_buff[2] = ports[port].transfer_bytes;
+        rb.Push(RESULT_SUCCESS.raw);
```
Comment (Rinkaa) [inline src/core/hle/service/cam/cam.cpp]: `rb.Push(RESULT_SUCCESS);`

File: src/core/hle/service/cam/cam.cpp [hunk 4]
```diff
@@ -958,35 +919,37 @@ void GetStereoCameraCalibrationData(Service::Interface* self) {
...
-    memcpy(&cmd_buff[2], &data, sizeof(data));
+    rb.Push(RESULT_SUCCESS);
+    rb.PushRaw(data);
 
     LOG_TRACE(Service_CAM, "called");
 }
 
 void SetPackageParameterWithoutContext(Service::Interface* self) {
-    u32* cmd_buff = Kernel::GetCommandBuffer();
+    IPC::RequestParser rp(Kernel::GetCommandBuffer(), 0x33, 11, 0);
```
Comment (Rinkaa) [inline src/core/hle/service/cam/cam.cpp]: Not something serious but I feel a bit confused with the `11` here, so maybe change it to a `const` ? 
Ignore this and next 3 comment items if the changes are unnessesary

File: src/core/hle/service/cam/cam.cpp [hunk 5]
```diff
@@ -958,35 +919,37 @@ void GetStereoCameraCalibrationData(Service::Interface* self) {
...
 
     PackageParameterWithoutContext package;
-    std::memcpy(&package, cmd_buff + 1, sizeof(package));
+    rp.PopRaw(package);
+    rp.Skip(4, false);
```
Comment (Rinkaa) [inline src/core/hle/service/cam/cam.cpp]: ditto ( `4`, is it `11`-`7`(ContextDetail) ?)

File: src/core/hle/service/cam/cam.cpp [hunk 6]
```diff
@@ -1019,36 +980,35 @@ Resolution PackageParameterWithContext::GetResolution() {
 }
 
 void SetPackageParameterWithContext(Service::Interface* self) {
-    SetPackageParameter<PackageParameterWithContext, 0x34>();
+    SetPackageParameter<PackageParameterWithContext, 0x34, 5>();
```
Comment (Rinkaa) [inline src/core/hle/service/cam/cam.cpp]: ditto ( `5` )

File: src/core/hle/service/cam/cam.cpp [hunk 7]
```diff
@@ -1019,36 +980,35 @@ Resolution PackageParameterWithContext::GetResolution() {
...
 }
 
 void SetPackageParameterWithContextDetail(Service::Interface* self) {
-    SetPackageParameter<PackageParameterWithContextDetail, 0x35>();
+    SetPackageParameter<PackageParameterWithContextDetail, 0x35, 7>();
```
Comment (Rinkaa) [inline src/core/hle/service/cam/cam.cpp]: ditto ( `7` )

File: src/core/hle/service/cam/cam.cpp (hunk 4 above)
Comment (wwylele) [inline src/core/hle/service/cam/cam.cpp]: CAM just uses these arbitrary param length in these functions, and I don't think use a named constant would help. As a convention elsewhere in our codebase, I just leave the number here.

**Comment by wwylele** [comment]
Addressed comments above. Thanks for the review!

**Comment by yuriks** [comment]
Reviewed 1 of 1 files at r1.
Review status: 0 of 1 files reviewed at latest revision, 15 unresolved discussions.

---

*[src/core/hle/service/cam/cam.cpp, line 322 at r1](https://reviewable.io:443/reviews/citra-emu/citra/2722#-Kl8ASzTVJowLaJgRk4E:-Kl8ASzUatq2iVrL9avO:btxz35i) ([raw file](https://github.com/citra-emu/citra/blob/924292f2cc7f7f766fdaa8d7d35bfe752b3f4bd1/src/core/hle/service/cam/cam.cpp#L322)):*
> ```cpp
>         LOG_ERROR(Service_CAM, "invalid port_select=%u", port_select.m_val);
>         rb.Push(ERROR_INVALID_ENUM_VALUE);
>         rb.Skip(1, false);
> ```

Why false (aka, not clear to 0)? (Also ugh this function is badly designed, the boolean means almost nothing on the callsite.)

---

*[src/core/hle/service/cam/cam.cpp, line 422 at r1](https://reviewable.io:443/reviews/citra-emu/citra/2722#-Kl8CE_rhfDKs7XmNxxi:-Kl8CE_sUOMts2pdC566:b6x7rv7) ([raw file](https://github.com/citra-emu/citra/blob/924292f2cc7f7f766fdaa8d7d35bfe752b3f4bd1/src/core/hle/service/cam/cam.cpp#L422)):*
> ```cpp
>         LOG_ERROR(Service_CAM, "invalid port_select=%u", port_select.m_val);
>         rb.Push(ERROR_INVALID_ENUM_VALUE);
>         rb.Skip(1, false);
> ```

Same, why false?

---

*[src/core/hle/service/cam/cam.cpp, line 581 at r1](https://reviewable.io:443/reviews/citra-emu/citra/2722#-Kl8EKqnItHOUfcMHjT5:-Kl8EKqnItHOUfcMHjT6:b85smdr) ([raw file](https://github.com/citra-emu/citra/blob/924292f2cc7f7f766fdaa8d7d35bfe752b3f4bd1/src/core/hle/service/cam/cam.cpp#L581)):*
> ```cpp
>         LOG_ERROR(Service_CAM, "invalid port_select=%u", port_select.m_val);
>         rb.Push(ERROR_INVALID_ENUM_VALUE);
>         rb.Skip(1, false);
> ```

the false again

---

*[src/core/hle/service/cam/cam.cpp, line 628 at r1](https://reviewable.io:443/reviews/citra-emu/citra/2722#-Kl8EX8-9OSfpv-m-8KT:-Kl8EX80A1FhkSPZOJ2Y:bqxf3j9) ([raw file](https://github.com/citra-emu/citra/blob/924292f2cc7f7f766fdaa8d7d35bfe752b3f4bd1/src/core/hle/service/cam/cam.cpp#L628)):*
> ```cpp
>         LOG_ERROR(Service_CAM, "invalid port_select=%u", port_select.m_val);
>         rb.Push(ERROR_INVALID_ENUM_VALUE);
>         rb.Skip(4, false);
> ```

I'll stop asking about this, make sure you change everything if you change it.

---

*[src/core/hle/service/cam/cam.cpp, line 948 at r1](https://reviewable.io:443/reviews/citra-emu/citra/2722#-Kl8IFduCmUqBdssTI80:-Kl8IFduCmUqBdssTI81:bcgajm5) ([raw file](https://github.com/citra-emu/citra/blob/924292f2cc7f7f766fdaa8d7d35bfe752b3f4bd1/src/core/hle/service/cam/cam.cpp#L948)):*
> ```cpp
>     PackageParameterType package;
>     rp.PopRaw(package);
>     rp.Skip(param_length - (sizeof(PackageParameterType) + 3) / 4, false);
> ```

Can we just take the reading out of this common template function? This is very ugly, doesn't de-duplicate significant shared code, and is very hard to understand what's going on. You can still pass the resulting package into this function so that it can set the camera settings, but the IPC-related stuff stays outside.

---

*[src/core/hle/service/cam/cam.cpp, line 993 at r1](https://reviewable.io:443/reviews/citra-emu/citra/2722#-Kl8IjbZInzf8FUo7fhK:-Kl8IjbZInzf8FUo7fhL:b-ye31to) ([raw file](https://github.com/citra-emu/citra/blob/924292f2cc7f7f766fdaa8d7d35bfe752b3f4bd1/src/core/hle/service/cam/cam.cpp#L993)):*
> ```cpp
> void GetSuitableY2rStandardCoefficient(Service::Interface* self) {
>     IPC::RequestBuilder rb =
>         IPC::RequestParser(Kernel::GetCommandBuffer(), 0x36, 0, 0).MakeBuilder(2, 0);
> ```

Rather split this in 2 statements than have the wrapper line, imo.

---

*[src/core/hle/service/cam/cam.cpp, line 1012 at r1](https://reviewable.io:443/reviews/citra-emu/citra/2722#-Kl8IoM4dqwU6HH0L__s:-Kl8IoM5tWPCPCj6i4BP:b-fx43j3) ([raw file](https://github.com/citra-emu/citra/blob/924292f2cc7f7f766fdaa8d7d35bfe752b3f4bd1/src/core/hle/service/cam/cam.cpp#L1012)):*
> ```cpp
> void DriverInitialize(Service::Interface* self) {
>     IPC::RequestBuilder rb =
>         IPC::RequestParser(Kernel::GetCommandBuffer(), 0x39, 0, 0).MakeBuilder(1, 0);
> ```

Same about splitting in 2 statements.

---

*[src/core/hle/service/cam/cam.cpp, line 1045 at r1](https://reviewable.io:443/reviews/citra-emu/citra/2722#-Kl8Isdj-tpncVUfAm1E:-Kl8Isdj-tpncVUfAm1F:bio604c) ([raw file](https://github.com/citra-emu/citra/blob/924292f2cc7f7f766fdaa8d7d35bfe752b3f4bd1/src/core/hle/service/cam/cam.cpp#L1045)):*
> ```cpp
> void DriverFinalize(Service::Interface* self) {
>     IPC::RequestBuilder rb =
>         IPC::RequestParser(Kernel::GetCommandBuffer(), 0x3A, 0, 0).MakeBuilder(1, 0);
> ```

And again.

---

*[src/core/hle/service/cam/cam.cpp, line 417 at r2](https://reviewable.io:443/reviews/citra-emu/citra/2722#-Kktu3h_nH5PrVakjf14-r2-417:-Kl8J1JDzKRTIbQOEjsB:b-pdfvrp) ([raw file](https://github.com/citra-emu/citra/blob/8cd9522526de290865094c7ff102db8e2872b471/src/core/hle/service/cam/cam.cpp#L417)):*
<details><summary><i>Previously, Rinkaa (RinkaaIkigai) wrote…</i></summary><blockquote>

Intentional?
</blockquote></details>

This is kinda hard to read like this, I think you should introduce a variable for this:
```c++
bool is_busy = ports[port].is_receiving || ports[port].is_pending_receiving;
rb.Push(!is_busy);
```

---

*[src/core/hle/service/cam/cam.cpp, line 933 at r2](https://reviewable.io:443/reviews/citra-emu/citra/2722#-Kktu3h_nH5PrVakjf14-r2-933:-Kl8H6Gnf6RYzmF2FQPq:bpy92x6) ([raw file](https://github.com/citra-emu/citra/blob/8cd9522526de290865094c7ff102db8e2872b471/src/core/hle/service/cam/cam.cpp#L933)):*
<details><summary><i>Previously, Rinkaa (RinkaaIkigai) wrote…</i></summary><blockquote>

ditto ( `4`, is it `11`-`7`(ContextDetail) ?)
</blockquote></details>

Maybe we can instead add padding to the struct so that its size matches what's popped off here. @wwylele Do you think that would make sense?

---

*[src/core/hle/service/cam/cam.cpp, line 517 at r3](https://reviewable.io:443/reviews/citra-emu/citra/2722#-Kl8Dx8epbU_ovOzHyjU:-Kl8Dx8epbU_ovOzHyjV:b-816h6l) ([raw file](https://github.com/citra-emu/citra/blob/8f02a1076567a6986e5bf3f61c2bb2f9bdee97b0/src/core/hle/service/cam/cam.cpp#L517)):*
> ```cpp
>         LOG_ERROR(Service_CAM, "invalid port_select=%u", port_select.m_val);
>         rb.Push(ERROR_INVALID_ENUM_VALUE);
>         rb.Skip(1, false);
> ```

Same.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/citra-emu/citra/2722)*

**Comment by wwylele** [comment]
Review status: 0 of 1 files reviewed at latest revision, 15 unresolved discussions.

---

*[src/core/hle/service/cam/cam.cpp, line 322 at r1](https://reviewable.io:443/reviews/citra-emu/citra/2722#-Kl8ASzTVJowLaJgRk4E:-KlA5B7M6DGGuwEKmZ3p:b-x7sgre) ([raw file](https://github.com/citra-emu/citra/blob/924292f2cc7f7f766fdaa8d7d35bfe752b3f4bd1/src/core/hle/service/cam/cam.cpp#L322)):*
<details><summary><i>Previously, yuriks (Yuri Kunde Schlesner) wrote…</i></summary><blockquote>

Why false (aka, not clear to 0)? (Also ugh this function is badly designed, the boolean means almost nothing on the callsite.)
</blockquote></details>

The previous implementation doesn't clear it to zero either, and the reason of it, is that most (if not all) camera service functions in real system module returns a uninitialized value to the command buffer when an error occur, which was also the reason I demand the `Skip` function.

---

*[src/core/hle/service/cam/cam.cpp, line 948 at r1](https://reviewable.io:443/reviews/citra-emu/citra/2722#-Kl8IFduCmUqBdssTI80:-KlA9hxYwEwf7S2Zu4BZ:b-896fix) ([raw file](https://github.com/citra-emu/citra/blob/924292f2cc7f7f766fdaa8d7d35bfe752b3f4bd1/src/core/hle/service/cam/cam.cpp#L948)):*
<details><summary><i>Previously, yuriks (Yuri Kunde Schlesner) wrote…</i></summary><blockquote>

Can we just take the reading out of this common template function? This is very ugly, doesn't de-duplicate significant shared code, and is very hard to understand what's going on. You can still pass the resulting package into this function so that it can set the camera settings, but the IPC-related stuff stays outside.
</blockquote></details>

Done.

---

*[src/core/hle/service/cam/cam.cpp, line 993 at r1](https://reviewable.io:443/reviews/citra-emu/citra/2722#-Kl8IjbZInzf8FUo7fhK:-KlA73j8FboP8fk-qu3I:b-896fix) ([raw file](https://github.com/citra-emu/citra/blob/924292f2cc7f7f766fdaa8d7d35bfe752b3f4bd1/src/core/hle/service/cam/cam.cpp#L993)):*
<details><summary><i>Previously, yuriks (Yuri Kunde Schlesner) wrote…</i></summary><blockquote>

Rather split this in 2 statements than have the wrapper line, imo.
</blockquote></details>

Done.

---

*[src/core/hle/service/cam/cam.cpp, line 1012 at r1](https://reviewable.io:443/reviews/citra-emu/citra/2722#-Kl8IoM4dqwU6HH0L__s:-KlA74Y_wwsBxSGKgPmX:b-896fix) ([raw file](https://github.com/citra-emu/citra/blob/924292f2cc7f7f766fdaa8d7d35bfe752b3f4bd1/src/core/hle/service/cam/cam.cpp#L1012)):*
<details><summary><i>Previously, yuriks (Yuri Kunde Schlesner) wrote…</i></summary><blockquote>

Same about splitting in 2 statements.
</blockquote></details>

Done.

---

*[src/core/hle/service/cam/cam.cpp, line 1045 at r1](https://reviewable.io:443/reviews/citra-emu/citra/2722#-Kl8Isdj-tpncVUfAm1E:-KlA75DyetMuunON87da:b-896fix) ([raw file](https://github.com/citra-emu/citra/blob/924292f2cc7f7f766fdaa8d7d35bfe752b3f4bd1/src/core/hle/service/cam/cam.cpp#L1045)):*
<details><summary><i>Previously, yuriks (Yuri Kunde Schlesner) wrote…</i></summary><blockquote>

And again.
</blockquote></details>

Done.

---

*[src/core/hle/service/cam/cam.cpp, line 417 at r2](https://reviewable.io:443/reviews/citra-emu/citra/2722#-Kktu3h_nH5PrVakjf14-r2-417:-KlA6Qkl13fbCJuieOC5:b-896fix) ([raw file](https://github.com/citra-emu/citra/blob/8cd9522526de290865094c7ff102db8e2872b471/src/core/hle/service/cam/cam.cpp#L417)):*
<details><summary><i>Previously, yuriks (Yuri Kunde Schlesner) wrote…</i></summary><blockquote>

This is kinda hard to read like this, I think you should introduce a variable for this:
```c++
bool is_busy = ports[port].is_receiving || ports[port].is_pending_receiving;
rb.Push(!is_busy);
```
</blockquote></details>

Done.

---

*[src/core/hle/service/cam/cam.cpp, line 933 at r2](https://reviewable.io:443/reviews/citra-emu/citra/2722#-Kktu3h_nH5PrVakjf14-r2-933:-KlA9k5UD4_SS0Ao9Cve:b-4ncwqd) ([raw file](https://github.com/citra-emu/citra/blob/8cd9522526de290865094c7ff102db8e2872b471/src/core/hle/service/cam/cam.cpp#L933)):*
<details><summary><i>Previously, yuriks (Yuri Kunde Schlesner) wrote…</i></summary><blockquote>

Maybe we can instead add padding to the struct so that its size matches what's popped off here. @wwylele Do you think that would make sense?
</blockquote></details>

good idea, done

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/citra-emu/citra/2722)*

**Comment by yuriks** [comment]
:lgtm:

---

Reviewed 2 of 2 files at r4.
Review status: all files reviewed at latest revision, 11 unresolved discussions.

---

*[src/core/hle/service/cam/cam.cpp, line 322 at r1](https://reviewable.io:443/reviews/citra-emu/citra/2722#-Kl8ASzTVJowLaJgRk4E:-KlC74Tk5fBQrwd1eZ02:b-mn28id) ([raw file](https://github.com/citra-emu/citra/blob/924292f2cc7f7f766fdaa8d7d35bfe752b3f4bd1/src/core/hle/service/cam/cam.cpp#L322)):*
<details><summary><i>Previously, wwylele (Weiyi Wang) wrote…</i></summary><blockquote>

The previous implementation doesn't clear it to zero either, and the reason of it, is that most (if not all) camera service functions in real system module returns a uninitialized value to the command buffer when an error occur, which was also the reason I demand the `Skip` function.
</blockquote></details>

I'm ok with doing this if the original request is larger than the response (so that what's returned it always just the bytes that came in from the request.) But using left-over garbage bytes from the buffer like in this case leads to too much non-determinism imo.

I'll probably end up making it so that unused parts of a request/response are cleared before/after calling the handler though, which would make this point moot. (The value would always be 0.)

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/citra-emu/citra/2722#-:-KlC8KJSVZnQiArhhG3I:bnfp4nl)*

**Pull request merged**